lib/vnet/dns/dns.go

```diff
@@ -0,0 +1,389 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package dns
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/net/dns/dnsmessage"
+	"golang.org/x/sync/errgroup"
+	"gvisor.dev/gvisor/pkg/tcpip"
+
+	"github.com/gravitational/teleport"
+)
+
+const (
+	maxUDPMessageSize = 65535
+)
+
+// Resolver represents an entity that can resolve DNS requests.
+type Resolver interface {
```

Comment:
I assume `net.Resolver` won't work here?

Reply:
right - `net.Resolver` is a concrete type that is an actual DNS resolver, I'm using this interface to pass in my custom DNS resolver in the tests and in https://github.com/gravitational/teleport/pull/41031